collect_app/src/test/java/org/odk/collect/android/widgets/WidgetFactoryTest.java

```diff
@@ -0,0 +1,345 @@
+import static org.mockito.Mockito.when;
+
+@RunWith(RobolectricTestRunner.class)
+public class WidgetFactoryTest {
+    private WidgetFactory widgetFactory;
+
+    @Before
+    public void setup() {
+        Activity activity = RobolectricHelpers.buildThemedActivity(TestScreenContextActivity.class).get();
+        WaitingForDataRegistry waitingForDataRegistry = new FakeWaitingForDataRegistry();
+        QuestionMediaManager questionMediaManager = new FakeQuestionMediaManager();
+        ActivityAvailability activityAvailability = mock(ActivityAvailability.class);
+
+        widgetFactory = new WidgetFactory(activity,
+                false,
+                false,
+                waitingForDataRegistry,
+                questionMediaManager,
+                null,
+                null,
+                activityAvailability,
+                null);
+    }
+
+    @Test
+    public void testCreatingSelectOneMinimalWidgetWithFastExternalChoices() {
+        QuestionDef questionDef = mock(QuestionDef.class);
+        when(questionDef.getAdditionalAttribute(null, "query")).thenReturn("instance('counties')/root/item[state= /data/state ]");
+
+        FormEntryPrompt prompt = new MockFormEntryPromptBuilder()
+                .withControlType(Constants.CONTROL_INPUT)
+                .withDataType(Constants.DATATYPE_TEXT)
+                .withQuestion(questionDef)
+                .withAppearance("search(‘items’) LisT miNimal something")
+                .build();
+
+        QuestionWidget widget = widgetFactory.createWidgetFromPrompt(prompt);
+        assertThat(widget, instanceOf(SelectOneMinimalWidget.class));
+    }
+
+    @Test
+    public void testCreatingLikertWidgetWithFastExternalChoices() {
+        QuestionDef questionDef = mock(QuestionDef.class);
+        when(questionDef.getAdditionalAttribute(null, "query")).thenReturn("instance('counties')/root/item[state= /data/state ]");
+
+        FormEntryPrompt prompt = new MockFormEntryPromptBuilder()
+                .withControlType(Constants.CONTROL_INPUT)
+                .withDataType(Constants.DATATYPE_TEXT)
+                .withQuestion(questionDef)
+                .withAppearance("search(‘items’) LisT lIkErt something")
+                .build();
+
+        QuestionWidget widget = widgetFactory.createWidgetFromPrompt(prompt);
+        assertThat(widget, instanceOf(LikertWidget.class));
+    }
+
+    @Test
+    public void testCreatingSelectOneListNoLabelWidgetWithFastExternalChoices() {
+        QuestionDef questionDef = mock(QuestionDef.class);
+        when(questionDef.getAdditionalAttribute(null, "query")).thenReturn("instance('counties')/root/item[state= /data/state ]");
+
+        FormEntryPrompt prompt = new MockFormEntryPromptBuilder()
+                .withControlType(Constants.CONTROL_INPUT)
+                .withDataType(Constants.DATATYPE_TEXT)
+                .withQuestion(questionDef)
+                .withAppearance("search(‘items’) LisT-nOLabeL something")
```

Comment:
These are really impractical appearance combinations. If you have a big enough dataset to user an external data file, why would you be building a grid?

Reply:
Same here this test is removed.